src/syft/core/node/common/plan/plan.py

```diff
@@ -0,0 +1,156 @@
+
+CAMEL_TO_SNAKE_PAT = re.compile(r"(?<!^)(?=[A-Z])")
+
+
+@bind_protobuf
+class Plan(Serializable):
+    """
+    A plan is a collection of actions, plus some variable inputs, that together form a computation graph.
+
+    Attributes:
+        actions: list of actions
+        inputs: Pointers to the inputs. Defaults to None.
+    """
+
+    def __init__(
+        self, actions: List[Action], inputs: Union[Pointer, List[Pointer], None] = None
+    ):
+        self.actions = actions
+        self.inputs: List[Pointer] = listify(inputs)
+
+    def execute(
+        self, node: AbstractNode, verify_key: VerifyKey, *args: Tuple[Any]
+    ) -> None:
+        """
+        1) For all pointers that were passed into the init as `inputs`, this method
+           replaces those pointers in self.actions by the pointers passed in as *args.
+        2) Executes the actions in self.actions one by one
+
+        *While this function requires `node` and `verify_key` as inputs, during remote
+        execution, passing these is handled in `RunClassMethodAction`*
+
+        *Note that this method will receive *args as pointers during execution. Normally,
+        pointers are resolved during `RunClassMethodAction.execute()`, but not for plans,
+        as they need to operate on the pointer to enable remapping of the inputs.*
+        Args:
+            *args: the new inputs for the plan, passed as pointers
+        """
+        inputs = listify(args)
+
+        # this is pretty cumbersome, we are searching through all actions to check
+        # if we need to redefine some of their attributes that are inputs in the
+        # graph of actions
+        for i, (current_input, new_input) in enumerate(zip(self.inputs, inputs)):
+            for a in self.actions:
+                if hasattr(a, "remap_input"):
+                    a.remap_input(current_input, new_input)
+
+            # redefine the inputs of the plan
+            self.inputs[i] = new_input
+
+        for a in self.actions:
+            a.execute_action(node, verify_key)
+
+    @staticmethod
+    def get_protobuf_schema() -> GeneratedProtocolMessageType:
+        """Return the type of protobuf object which stores a class of this type
+
+        As a part of serialization and deserialization, we need the ability to
+        lookup the protobuf object type directly from the object type. This
+        static method allows us to do this.
+
+        Importantly, this method is also used to create the reverse lookup ability within
+        the metaclass of Serializable. In the metaclass, it calls this method and then
+        it takes whatever type is returned from this method and adds an attribute to it
+        with the type of this class attached to it. See the MetaSerializable class for details.
+
+        :return: the type of protobuf object which corresponds to this class.
+        :rtype: GeneratedProtocolMessageType
+
+        """
+
+        return Plan_PB
+
+    def _object2proto(self) -> Plan_PB:
```

Comment:
If you merge `dev` you can update to the new Storable interface which removes lots of boiler plate.